### PR TITLE
new_bls12_hash_function-milestone_2

### PR DESCRIPTION
**Pull request by dishport**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [ ] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/966

### CODE REVIEW

File: deliveries/new_bls12_hash_function-milestone_1.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Milestone Delivery :mailbox:
+
+**The [invoice form :pencil:](https://docs.google.com/forms/d/e/1FAIpQLSfmNYaoCgrxyhzgoKQ0ynQvnNRoTmgApz9NrMp-hd8mhIiO0A/viewform) has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).**  
+
+* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/new_bls12_hash_function.md
+* **Milestone Number:** 1
```
Comment (alxs) [inline deliveries/new_bls12_hash_function-milestone_1.md:6]:
```suggestion
* **Milestone Number:** 2
```

**Comment by alxs** [comment]
Thanks for the new delivery @dishport. @drskalman did a great review of your first milestone and is now familiar with your function, so he'll review this delivery as well.

Could you please fill in the checkmarks in the comment template above? You can already submit a separate invoice for this milestone with the previous one.

File: deliveries/new_bls12_hash_function-milestone_2.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+* **Application Document:** https://github.com/w3f/Grants-Program/blob/master/applications/new_bls12_hash_function.md
+* **Milestone Number:** 2
+
+**Context** (optional)
+
+**Deliverables**
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
+| 0a. | License | Apache License, Version 2.0 (https://www.apache.org/licenses/LICENSE-2.0) and MIT license (https://opensource.org/licenses/MIT) | ... |
```
Comment (alxs) [inline deliveries/new_bls12_hash_function-milestone_2.md:14]: Please add a [license file](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/adding-a-license-to-a-repository#including-an-open-source-license-in-your-repository) to the repository

File: deliveries/new_bls12_hash_function-milestone_2.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+| 0b. | Documentation | published article: https://link.springer.com/article/10.1007/s10623-022-01012-8, free eprint version: https://eprint.iacr.org/2021/301 | ... |
```
Comment (alxs) [inline deliveries/new_bls12_hash_function-milestone_2.md:15]: This deliverable refers to inline documentation
```suggestion
| 0b. | Documentation |  | Inline documentation |
```

File: deliveries/new_bls12_hash_function-milestone_2.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+| 0c. | Testing Guide | make test | You may need [SageMath](https://www.sagemath.org/) to validate test vectors
```
Comment (alxs) [inline deliveries/new_bls12_hash_function-milestone_2.md:16]:
```suggestion
| 0c. | Testing Guide | https://github.com/zhenfeizhang/indifferentiable-hashing#testing | You may need [SageMath](https://www.sagemath.org/) to validate test vectors
```

**Review by alxs — CHANGES_REQUESTED** [review]

**Comment by dishport** [comment]
> Could you please fill in the checkmarks in the comment template above? You can already submit a separate invoice for this milestone with the previous one.

@alxs, Who should fill in the invoice form for this milestone: me or @zhenfeizhang ? In other words, will you send the money directly to him or through me ?

File: deliveries/new_bls12_hash_function-milestone_2.md (hunk 2 above)
Comment (dishport) [reply deliveries/new_bls12_hash_function-milestone_2.md:14]: @alxs, could you give a link to this repository ? I cannot find the necessary page on GitHub.

Comment (alxs) [reply deliveries/new_bls12_hash_function-milestone_2.md:14]: https://github.com/zhenfeizhang/indifferentiable-hashing

**Comment by alxs** [comment]
@dishport that would be you. You'll need to submit an invoice and payments should go to the same entity for one grant.

File: deliveries/new_bls12_hash_function-milestone_2.md (hunk 2 above)
Comment (zhenfeizhang) [reply deliveries/new_bls12_hash_function-milestone_2.md:14]: added an MIT license: https://github.com/zhenfeizhang/indifferentiable-hashing/blob/master/LICENSE

Comment (alxs) [reply deliveries/new_bls12_hash_function-milestone_2.md:14]:
```suggestion
| 0a. | License | [LICENSE](https://github.com/zhenfeizhang/indifferentiable-hashing/blob/master/LICENSE) | MIT |
```

**Comment by drskalman** [comment]
@zhenfeizhang Please resolve https://github.com/zhenfeizhang/indifferentiable-hashing/issues/1 so I am able to compile the code. thanx!

**Comment by zhenfeizhang** [comment]
> @zhenfeizhang Please resolve [zhenfeizhang/indifferentiable-hashing#1](https://github.com/zhenfeizhang/indifferentiable-hashing/issues/1) so I am able to compile the code. thanx!

Just pushed a fix. Please try again.

**Comment by drskalman** [comment]
@zhenfeizhang please resolve [issue 2](https://github.com/zhenfeizhang/indifferentiable-hashing/issues/2) and [issue 3](https://github.com/zhenfeizhang/indifferentiable-hashing/issues/3)

**Comment by alxs** [comment]
@zhenfeizhang @dishport friendly reminder

**Comment by zhenfeizhang** [comment]
addressed both issues. Please take a look @alxs @drskalman

**Comment by zhenfeizhang** [comment]
ping on this -- the fixes are relatively small so can we please have a quick review?

**Comment by alxs** [comment]
Hey @zhenfeizhang, we're on retreat this week so I kindly ask you for some (more) patience. @drskalman is also not officially in the grants team and only has limited time to review the delivery, but I pinged him about it. Thank you for your understanding.

**Comment by drskalman** [comment]
@zhenfeizhang https://github.com/zhenfeizhang/indifferentiable-hashing/issues/4

**Comment by dishport** [comment]
> @zhenfeizhang [zhenfeizhang/indifferentiable-hashing#4](https://github.com/zhenfeizhang/indifferentiable-hashing/issues/4)

@drskalman, could you check the corrected code of @zhenfeizhang ?

**Comment by drskalman** [comment]
The tests still failing for me for BLS12-381. I commented on the issue.

**Comment by Noc2** [comment]
@dishport What is the status here? Could you take a look at these issues: https://github.com/zhenfeizhang/indifferentiable-hashing/issues?

**Review by Noc2 — CHANGES_REQUESTED** [review]
(see message above)

**Comment by zhenfeizhang** [comment]
I have resolved the bug in https://github.com/zhenfeizhang/indifferentiable-hashing/issues/4
An integrated version for arkworks 0.4.0 is available, see https://github.com/zhenfeizhang/indifferentiable-hashing/issues/3#issuecomment-1407238397

**Review by drskalman — APPROVED** [review]
LGTM: https://github.com/w3f/Grant-Milestone-Delivery/pull/720

**Review by Noc2 — APPROVED** [review]
I'm happy to tell you that the milestone is finally approved. See https://github.com/w3f/Grant-Milestone-Delivery/pull/720 I will forward your invoice internally.

**Comment by RouvenP** [comment]
hi @drskalman we transferred the payment yesterday.

**Comment by dishport** [comment]
> hi @drskalman we transferred the payment yesterday.

Hi @RouvenP. I received. Thanks.

**Pull request merged**